File: 정수빈/LeetCode/medium/word-search.py
# https://leetcode.com/problems/word-search/?envType=problem-list-v2&envId=backtracking
# 설명: bk

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        n = len(board)
        m = len(board[0])
        global ans
        ans = False

        
        def bk(idx, r, c):
            global ans
            if idx == len(word)-1:
                ans = True
                return 
            
            for dr, dc in zip([0, 0, 1, -1], [1, -1, 0, 0]):
                nr, nc = r+dr, c+dc
                if 0<= nr < n and 0<= nc < m and not (nr, nc) in visited and word[idx+1] == board[nr][nc]:
                    visited.append((nr, nc))
                    bk(idx+1, nr, nc)
                    visited.pop()
        
        for r in range(n):
            for c in range(m):
                if board[r][c] == word[0]:
                    visited = [(r, c)]
                    bk(0, r, c)
                if ans:
                    break
        return ans

File: 장세훈/LeetCode/medium/WordSearch.java
class WordSearch {

  /*
      https://leetcode.com/problems/word-search/
   */

  boolean result = false;

  public boolean exist(char[][] board, String word) {
    for (int i = 0; i < board.length; i++) {
      for (int j = 0; j < board[i].length; j++) {
        if (board[i][j] == word.charAt(0)) {
          boolean[][] visit = new boolean[board.length][board[0].length];
          visit[i][j] = true;
          dfs(board, word, visit, i, j, 0);
        }
      }
    }

    return result;
  }


  void dfs(char[][] board, String word, boolean[][] visit, int x, int y, int idx) {
    if (result) {
      return;
    }

    if (word.length() == (idx + 1)) {
      result = true;
      return;
    }

    int[][] dir = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    for (int[] d : dir) {
      int nx = x + d[0];
      int ny = y + d[1];

      if (nx > -1 && ny > -1 && nx < board.length && ny < board[0].length && !visit[nx][ny] && idx + 1 < word.length()
          && board[nx][ny] == word.charAt(idx + 1)) {
        visit[nx][ny] = true;
        dfs(board, word, visit, nx, ny, idx + 1);
        visit[nx][ny] = false;
      }
    }
  }

}